### PR TITLE
[receiver/sqlqueryreceiver] Use one connection pool per receiver

### LINKED ISSUE #39270: [receiver/sqlquery] Multiple connections despite max_open_conn = 1
### Component(s)

receiver/sqlquery

## Description

#36752 is happening despite #37748. The sqlquery receiver opens one connection for every query, regardless of the `max_open_conn` setting.

## Steps to Reproduce

Use the configuration below with a recent collector and Postgres:

```shell
$ docker run --rm -it --net host --env POSTGRES_HOST_AUTH_METHOD=trust docker.io/library/postgres

$ docker run --rm -it --net host -v "$(pwd)/config.yaml:/etc/otelcol-contrib/config.yaml:ro" \
ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib:0.123.1
```

Use `psql` to see the open connections at Postgres:

```console
$ psql -h localhost -U postgres -c 'SELECT pid, backend_type, datname, usename, query, state FROM pg_stat_activity'
┌─────┬──────────────────────────────┬──────────┬──────────┬────────────────────────────────────────────────────────────────────────────────┬────────┐
│ pid │         backend_type         │ datname  │ usename  │                                     query                                      │ state  │
├─────┼──────────────────────────────┼──────────┼──────────┼────────────────────────────────────────────────────────────────────────────────┼────────┤
│ 102 │ client backend               │ postgres │ postgres │ SELECT 1 AS number                                                             │ idle   │
│ 103 │ client backend               │ postgres │ postgres │ SELECT 2 AS number                                                             │ idle   │
│ 104 │ client backend               │ postgres │ postgres │ SELECT 3 AS number                                                             │ idle   │
│ 110 │ client backend               │ postgres │ postgres │ SELECT pid, backend_type, datname, usename, query, state FROM pg_stat_activity │ active │
│  65 │ autovacuum launcher          │ ␀        │ ␀        │                                                                                │ ␀      │
│  66 │ logical replication launcher │ ␀        │ postgres │                                                                                │ ␀      │
│  61 │ checkpointer                 │ ␀        │ ␀        │                                                                                │ ␀      │
│  62 │ background writer            │ ␀        │ ␀        │                                                                                │ ␀      │
│  64 │ walwriter                    │ ␀        │ ␀        │                                                                                │ ␀      │
└─────┴──────────────────────────────┴──────────┴──────────┴────────────────────────────────────────────────────────────────────────────────┴────────┘
(9 rows)
```

## Expected Result

I expected one "client backend" connection because `max_open_conn: 1` in the receiver config.

## Actual Result

Three "client backend" connections, one for each query.

### Collector version

v0.123.1

## Environment
OS: Debian 12 (bookworm)

### OpenTelemetry Collector configuration

```yaml
---
receivers:
  sqlquery:
    driver: postgres
    datasource: host=localhost sslmode=disable user=postgres
    max_open_conn: 1
    collection_interval: 10s
    queries:
      - sql: 'SELECT 1 AS number'
        metrics:
          - metric_name: num_1
            value_column: number
      - sql: 'SELECT 2 AS number'
        metrics:
          - metric_name: num_2
            value_column: number
      - sql: 'SELECT 3 AS number'
        metrics:
          - metric_name: num_3
            value_column: number
exporters:
  debug: {}
service:
  pipelines:
    metrics:
      receivers: [sqlquery]
      processors: []
      exporters: [debug]
```

### Log output

```shell
2025-04-09T04:03:08.577Z        info    service@v0.123.0/service.go:197 Setting up own telemetry...
2025-04-09T04:03:08.578Z        info    builders/builders.go:26 Development component. May change in the future.
2025-04-09T04:03:08.579Z        info    service@v0.123.0/service.go:264 Starting otelcol-contrib...     {"Version": "0.123.1", "NumCPU": 12}
2025-04-09T04:03:08.579Z        info    extensions/extensions.go:41     Starting extensions...
2025-04-09T04:03:08.579Z        info    service@v0.123.0/service.go:287 Everything is ready. Begin running and processing data.
2025-04-09T04:03:09.600Z        info    Metrics {"resource metrics": 3, "metrics": 3, "data points": 3}
2025-04-09T04:03:19.583Z        info    Metrics {"resource metrics": 3, "metrics": 3, "data points": 3}
2025-04-09T04:03:29.583Z        info    Metrics {"resource metrics": 3, "metrics": 3, "data points": 3}
2025-04-09T04:03:39.583Z        info    Metrics {"resource metrics": 3, "metrics": 3, "data points": 3}
```

### Additional context

This function is called during every Scraper's startup, creating multiple pools:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/19d98da603a532a8b5f11f8a340746dac3b2ad44/receiver/sqlqueryreceiver/receiver.go#L47-L48

### PR DESCRIPTION
**Pull request by cbandy**
#### Description
Rather than creating a separate `*sql.DB` provider function for each query, this creates one provider for the receiver and passes a method value to each query. The method uses a mutex to ensure exactly one `*sql.DB` is created when it is called concurrently.

The `max_open_conn` config is defined in `sqlqueryreceiver`, so I put the implementation in its `internal` package.

#### Link to tracking issue
Fixes #39270

#### Testing
Using the configuration and steps in #39270, I see only one connection to Postgres:

```console
$ psql -h localhost -U postgres -c 'SELECT pid, backend_type, datname, usename, query, state FROM pg_stat_activity'
┌─────┬──────────────────────────────┬──────────┬──────────┬────────────────────────────────────────────────────────────────────────────────┬────────┐
│ pid │         backend_type         │ datname  │ usename  │                                     query                                      │ state  │
├─────┼──────────────────────────────┼──────────┼──────────┼────────────────────────────────────────────────────────────────────────────────┼────────┤
│  68 │ client backend               │ postgres │ postgres │ SELECT 3 AS number                                                             │ idle   │
│  70 │ client backend               │ postgres │ postgres │ SELECT pid, backend_type, datname, usename, query, state FROM pg_stat_activity │ active │
│  65 │ autovacuum launcher          │ ␀        │ ␀        │                                                                                │ ␀      │
│  66 │ logical replication launcher │ ␀        │ postgres │                                                                                │ ␀      │
│  61 │ checkpointer                 │ ␀        │ ␀        │                                                                                │ ␀      │
│  62 │ background writer            │ ␀        │ ␀        │                                                                                │ ␀      │
│  64 │ walwriter                    │ ␀        │ ␀        │                                                                                │ ␀      │
└─────┴──────────────────────────────┴──────────┴──────────┴────────────────────────────────────────────────────────────────────────────────┴────────┘
(7 rows)
```

### CODE REVIEW

File: receiver/sqlqueryreceiver/internal/database_sql.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// DB initializes and returns the [sql.DB] pool. It is safe to call concurrently.
+// Depending on the driver, this might also connect to the database backend.
+//
+// This method exists to satisfy [sqlquery.DbProviderFunc], but the way
+// [sqlquery.Scraper] closes [sql.DB] can interfere with other Scrapers.
+func (sp *sqlPool) DB() (*sql.DB, error) {
+	sp.mutex.Lock()
+	defer sp.mutex.Unlock()
```
Comment (crobert-1) [inline receiver/sqlqueryreceiver/internal/database_sql.go:41]: I'm concerned about some edge case where this mutex hangs due to the `sp.Opener` call. Have this been tested with an invalid database configuration, or do you have some other idea for how to test?

Comment (cbandy) [reply receiver/sqlqueryreceiver/internal/database_sql.go:41]: I understand your concern. You're worried that a slow Open will prevent this receiver's scrapers from scraping. That's how it goes if we want to share the pool among the scrapers.

Are you worried that this might affect startup/configuration? This function is called after config, when the receiver starts.

I can mock an opener that returns an error. 🤔 That would release the mutex and bubble up during Start. A panic here would do the same.

File: receiver/sqlqueryreceiver/receiver.go [hunk 2]
```diff
@@ -39,29 +39,22 @@ func createMetricsReceiverFunc(sqlOpenerFunc sqlquery.SQLOpenerFunc, clientProvi
...
 		sqlCfg := cfg.(*Config)
 		var opts []scraperhelper.ControllerOption
+		pool := internal.NewPool(sqlOpenerFunc, sqlCfg.Driver, sqlCfg.DataSource, sqlCfg.MaxOpenConn)
+
 		for i, query := range sqlCfg.Queries {
 			if len(query.Metrics) == 0 {
 				continue
 			}
 			id := component.MustNewIDWithName("sqlqueryreceiver", fmt.Sprintf("query-%d: %s", i, query.SQL))
-			dbProviderFunc := func() (*sql.DB, error) {
```
Comment (cbandy) [inline receiver/sqlqueryreceiver/receiver.go:47]: 📝 I can do away with the type, put a mutex on L42, and use it here in this provider func (closure).

File: receiver/sqlqueryreceiver/internal/database_sql.go (hunk 1 above)
Comment (crobert-1) [reply receiver/sqlqueryreceiver/internal/database_sql.go:41]: It might be alright for now, we can revisit this if we end up running into any issues. 👍

**Review by crobert-1 — APPROVED** [review]
